### PR TITLE
1810 Move mistral evaluation

### PR DESCRIPTION
**Pull request by Yousof-kayal**
#### Context
tracker: [#1810 ](https://github.com/pytorch/torchtune/issues/1810)
What is the purpose of this PR? Is it to
- [ ] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [x] clean up

Please link to any issues this PR addresses.

#### Changelog
What are the changes made in this PR?
* Copied `evaluation.yaml` to /recipes/mistral/ directory
* Updated `evaluation.yaml` to point to mistral 7b model instantiations
* Updated the recipe registry to pick up new configuration

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [x] manually run any new or modified recipes with sufficient proof of correctness
- [x] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### **Mistral7B v0.1 Eleuther evaluation recipe output:**
```
batch_size: 2
checkpointer:
  _component_: torchtune.training.FullModelHFCheckpointer
  checkpoint_dir: /tmp/Mistral-7B-v0.1/
  checkpoint_files:
  - pytorch_model-00001-of-00002.bin
  - pytorch_model-00002-of-00002.bin
  model_type: MISTRAL
  output_dir: /tmp/Mistral-7B-v0.1/
  recipe_checkpoint: null
device: cuda
dtype: bf16
enable_kv_cache: true
limit: null
max_seq_length: 4096
model:
  _component_: torchtune.models.mistral.mistral_7b
quantizer: null
resume_from_checkpoint: false
seed: 1234
tasks:
- truthfulqa_mc2
tokenizer:
  _component_: torchtune.models.mistral.mistral_tokenizer
  max_seq_len: null
  path: /tmp/Mistral-7B-v0.1/tokenizer.model

Model is initialized with precision torch.bfloat16.
2024-10-14:16:00:47,974 INFO     [eleuther_eval.py:524] Model is initialized with precision torch.bfloat16.
2024-10-14:16:00:48,413 INFO     [huggingface.py:129] Using device 'cuda:0'
2024-10-14:16:00:49,036 INFO     [huggingface.py:481] Using model type 'default'
2024-10-14:16:00:49,329 INFO     [huggingface.py:365] Model parallel was set to False, max memory was not set, and device map was set to {'': 'cuda:0'}
2024-10-14:16:00:55,159 INFO     [__init__.py:459] The tag 'arc_ca' is already registered as a group, this tag will not be registered. This may affect tasks you want to call.
2024-10-14:16:00:55,161 INFO     [__init__.py:459] The tag 'arc_ca' is already registered as a group, this tag will not be registered. This may affect tasks you want to call.
Running evaluation on the following tasks: ['truthfulqa_mc2']
2024-10-14:16:01:05,870 INFO     [eleuther_eval.py:568] Running evaluation on the following tasks: ['truthfulqa_mc2']
2024-10-14:16:01:05,871 INFO     [task.py:415] Building contexts for truthfulqa_mc2 on rank 0...
100%|███████████████████████████████████████████████████████████████████████████████| 817/817 [00:00<00:00, 1301.01it/s]
2024-10-14:16:01:06,535 INFO     [evaluator.py:489] Running loglikelihood requests
Running loglikelihood requests: 100%|█████████████████████████████████████████████| 5882/5882 [5:03:14<00:00,  3.09s/it]
Eval completed in 18197.52 seconds.
2024-10-14:21:04:23,386 INFO     [eleuther_eval.py:577] Eval completed in 18197.52 seconds.
Max memory allocated: 22.34 GB
2024-10-14:21:04:23,388 INFO     [eleuther_eval.py:578] Max memory allocated: 22.34 GB


|    Tasks     |Version|Filter|n-shot|Metric|   |Value |   |Stderr|
|--------------|------:|------|-----:|------|---|-----:|---|-----:|
|truthfulqa_mc2|      2|none  |     0|acc   |↑  |0.4268|±  |0.0142|

### CODE REVIEW

**Comment by joecummings** [comment]
@Yousof-kayal Can you please rebase/merge the main branch?

File: .gitignore [hunk 1]
```diff
@@ -53,6 +53,7 @@ venv/
 ENV/
 env.bak/
 venv.bak/
+torchvenv/
```
Comment (joecummings) [inline .gitignore:56]: We shouldn't add this here

File: recipes/configs/mistral/evaluation.yaml [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+# Checkpointer
+checkpointer:
+  _component_: torchtune.training.FullModelHFCheckpointer
+  checkpoint_dir: /tmp/Mistral-7B-v0.1/
+  checkpoint_files: [
+    pytorch_model-00001-of-00002.bin,
+    pytorch_model-00002-of-00002.bin
+  ]
+  recipe_checkpoint: null
```
Comment (joecummings) [inline recipes/configs/mistral/evaluation.yaml:18]: Can you remove this key? It's not needed.

**Review by joecummings — COMMENTED** [review]
Looks pretty good! Just a couple of nits.

File: recipes/configs/mistral/evaluation.yaml (hunk 2 above)
Comment (Yousof-kayal) [reply recipes/configs/mistral/evaluation.yaml]: Removed, thanks!

File: .gitignore (hunk 1 above)
Comment (Yousof-kayal) [reply .gitignore]: Removed, Thanks!

File: .gitignore [hunk 3]
```diff
@@ -186,4 +186,4 @@ cover/
 .vscode/
 
 # wandb
-wandb/
+wandb/
```
Comment (joecummings) [inline .gitignore:189]: Let's remove all modifications from this file?

Comment (Yousof-kayal) [reply .gitignore:189]: How do I revert it?
I tried removing the extra line but it's still changed.

Comment (joecummings) [reply .gitignore:189]: There's a couple possible ways:
1. I believe you can do this through the github UI by selecting `Discard Changes`:
Screenshot 2024-10-14 at 3 56 06 PM
2. Reset the commits and don't add the changes to the file (from git CLI): https://git-scm.com/book/en/v2/Git-Basics-Undoing-Things
3. Copy and paste the file from master,  which hopefully shouldn't have the changes.

**Review by joecummings — APPROVED** [review]
Thanks for your patience with this one @Yousof-kayal! Looks great, I'll merge when CI passes

**Comment by Yousof-kayal** [comment]
Don't worry about it!
